### PR TITLE
Introduced tree structure for projects and project images

### PR DESCRIPTION
**Pull request by gertjanal**
Added /projects and /projects/{projectId}/images REST calls.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -6,19 +6,6 @@ Install via NPM:
 
 ```bash
 npm install hansken-js
-
 ```
 
-
```
Comment (ghost) [inline README.md:12]: I kinda like having a simple code block in the readme. So maybe keep this here and add the `node examples/app.js` for more examples?

File: src/hansken.js [hunk 2]
```diff
@@ -1,13 +1,22 @@
...
-import { Projects } from './modules/projects.js';
+import { ProjectContext } from './modules/projectContext.js';
 
 class HanskenClient {
     
     constructor(gatekeeperUrl) {
-      const sessionManager = new SessionManager(gatekeeperUrl);
-      this.projects = new Projects(sessionManager);
+      this.sessionManager = new SessionManager(gatekeeperUrl);
+
```
Comment (ghost) [inline src/hansken.js]: legel

File: src/modules/projectContext.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+     * Get the project.
+     *
+     * @returns The project
+     */
+    get = () => this.sessionManager.gatekeeper(`/projects/${this.projectId}`).then((response) => response.json());
+
+    /**
+     * Update an existing project.
+     *
+     * @param {string} projectId The uuid of the project to update
```
Comment (ghost) [inline src/modules/projectContext.js]: This is not actually a parameter a user gives since the API is just a project.

`client.project('abcd').update({...});`

File: src/modules/projectContext.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        headers: {
+        'Content-Type': 'application/json'
+        },
+        body: JSON.stringify(project)
+    });
+
+    /**
+     * Get all images linked to a specific project.
+     *
+     * @param {string} projectId The uuid of the project to update
```
Comment (ghost) [inline src/modules/projectContext.js]: again, the projectId does not seem to be an actual parameter

File: src/modules/projectImageContext.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        this.projectId = projectId;
+        this.imageId = imageId;
+    }
+
+    /**
+     * Get the project image.
+     *
+     * @returns The project image
+     */
+    get = () => this.sessionManager.gatekeeper(`/projects/${this.projectId}/images/${this.imageId}`).then((response) => response.json());
```
Comment (ghost) [inline src/modules/projectImageContext.js:21]: Maybe to align the sessionManager weith the naming scheme, call it `HanskenContext`?

`this.hanskenContext.gatekeeper(...);`

File: src/modules/projectContext.js (hunk 3 above)
Comment (gertjanal) [reply src/modules/projectContext.js]: True, outdated doc

File: src/hansken.js (hunk 2 above)
Comment (gertjanal) [reply src/hansken.js]: Removed

File: src/modules/projectImageContext.js (hunk 5 above)
Comment (gertjanal) [reply src/modules/projectImageContext.js:21]: Discussed IRL. The user doesn't see this object, and the only goal of the object is to keep a valid session with any of the Hansken services.

File: README.md (hunk 1 above)
Comment (gertjanal) [reply README.md:12]: Reintroduced

File: README.md [hunk 6]
```diff
@@ -6,19 +6,19 @@ Install via NPM:
...
-<script type="text/javascript" src="hansken.js"></script>
-<script type="text/javascript">
-    const hansken = new HanskenClient('https://gatekeeper01.hansken.org/gatekeeper');
+import {HanskenClient} from './src/hansken.js';
 
-    hansken.projects().then((projects) => {
-        projects.forEach((project) => {
-            document.write(`<li>${project.name}`);
-        });
+const hansken = new HanskenClient('/gatekeeper');
```
Comment (ghost) [inline README.md:15]: now that I see the hansken client, maybe you can make this an object? like so;
```javascript
const config = {
  gatekeeper: '/gatekeeper',
  preference: '/preference'
}

const hansken = new HanskenClient(config);
```

File: src/modules/projectContext.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
+import { ProjectImageContext } from './projectImageContext.js';
+
+class ProjectContext {
+
+    constructor(sessionManager, projectId) {
```
Comment (ghost) [inline src/modules/projectContext.js:11]: javascript doc

Comment (ghost) [reply src/modules/projectContext.js:11]: maybe also specify projectId has to be a uuid

File: src/modules/projectContext.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    });
+
+    /**
+     * Get all images linked to a specific project.
+     *
+     * @returns An array of images linked to this project
+     */
+    images = () => this.sessionManager.gatekeeper(`/projects/${this.projectId}/images`).then((response) => response.json());
+
+    image = (imageId) => new ProjectImageContext(this.sessionManager, this.projectId, imageId);
```
Comment (ghost) [inline src/modules/projectContext.js:50]: javascript doc

File: src/hansken.js [hunk 9]
```diff
@@ -1,13 +1,21 @@
 import { SessionManager } from './modules/sessionManager.js';
-import { Projects } from './modules/projects.js';
+import { ProjectContext } from './modules/projectContext.js';
 
 class HanskenClient {
     
     constructor(gatekeeperUrl) {
```
Comment (ghost) [inline src/hansken.js:11]: javascript doc

File: src/hansken.js [hunk 10]
```diff
@@ -1,13 +1,21 @@
...
     }
 
+    /**
+     * Get all projects.
+     *
+     * @returns All projects the current user is authorized for
+     */
+    projects = () => this.sessionManager.gatekeeper('/projects').then((response) => response.json());
+
+    project = (projectId) => new ProjectContext(this.sessionManager, projectId);
```
Comment (ghost) [inline src/hansken.js]: javascript doc

File: src/modules/sessionManager.js [hunk 11]
```diff
@@ -58,7 +58,7 @@ class SessionManager {
             });
     };
 
-    gatekeeper = (url, req) => SessionManager.#fetch(this.gatekeeperUrl, url, req);
+    gatekeeper = (path, request) => SessionManager.#fetch(this.gatekeeperUrl, path, request);
```
Comment (ghost) [inline src/modules/sessionManager.js:73]: javascript doc

**Review by ghost — APPROVED** [review]

File: examples/project-images.html [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+    const projectId = params.projectId;
+    if (projectId) {
+        const hansken = new HanskenClient('/gatekeeper');
+        hansken.project(projectId).images().then((images) => {
+            images.forEach((image) => {
+                document.write(`<li>${image.description}`);
+            });
+        });
+    }
+</script>
```
Comment (dijkstraj) [inline examples/project-images.html]: No newline at end of file here, but some other files do. To help avoid unnecessary changes this should be consistent for all files (maybe add a `.editorconfig`?).

File: examples/projects.html [hunk 13]
```diff
@@ -0,0 +1,10 @@
+<script type="module">
+    import {HanskenClient} from './src/hansken.js';
+   
+    const hansken = new HanskenClient('/gatekeeper');
+    hansken.projects().then((projects) => {
+        projects.forEach((project) => {
+            document.write(`<li><a href="project-images.html?projectId=${project.id}">${project.name}</a>`);
+        });
+    });
+</script>
```
Comment (dijkstraj) [inline examples/projects.html]: No newline, see previous comment.

File: package.json [hunk 14]
```diff
@@ -22,5 +22,8 @@
     "type": "git",
     "url": "https://github.com/NetherlandsForensicInstitute/hansken-js.git"
   },
-  "devDependencies": {}
+  "devDependencies": {},
+  "dependencies": {
+    "express": "^4.17.1"
```
Comment (dijkstraj) [inline package.json:26]: Shouldn't this be a dev dependency? Looks like it's only used to run examples.

File: README.md (hunk 6 above)
Comment (gertjanal) [reply README.md:15]: I expect that this constructor will take no more than 3 arguments in the future; gatekeeper, keystore and preference urls. For now, I'd like to keep it this way as it takes less code to initialize. When we need more config parameters, we can add your suggestion in a future release.

File: src/modules/projectContext.js (hunk 7 above)
Comment (gertjanal) [reply src/modules/projectContext.js:11]: Added

File: src/modules/projectContext.js (hunk 8 above)
Comment (gertjanal) [reply src/modules/projectContext.js:50]: Added

File: src/hansken.js (hunk 9 above)
Comment (gertjanal) [reply src/hansken.js:11]: Added

File: src/hansken.js (hunk 10 above)
Comment (gertjanal) [reply src/hansken.js]: Added

File: src/modules/sessionManager.js (hunk 11 above)
Comment (gertjanal) [reply src/modules/sessionManager.js:73]: Added

File: examples/project-images.html (hunk 12 above)
Comment (gertjanal) [reply examples/project-images.html]: Added new lines

File: examples/projects.html (hunk 13 above)
Comment (gertjanal) [reply examples/projects.html]: Added

File: package.json (hunk 14 above)
Comment (gertjanal) [reply package.json:26]: True

Comment (gertjanal) [reply package.json:26]: MOved

File: src/modules/sessionManager.js [hunk 15]
```diff
@@ -1,5 +1,10 @@
 class SessionManager {
 
+    /**
+     * Creates an object that handles the authentication of SAML services.
+     *
+     * @param {String} gatekeeperUrl The url to the Hansken gatekeeper, without trailing '/'
```
Comment (dijkstraj) [inline src/modules/sessionManager.js]:
```suggestion
     * @param {string} gatekeeperUrl The url to the Hansken gatekeeper, without trailing '/'
```
It's lowercase in js

File: src/modules/sessionManager.js [hunk 16]
```diff
@@ -58,7 +63,14 @@ class SessionManager {
             });
     };
 
-    gatekeeper = (url, req) => SessionManager.#fetch(this.gatekeeperUrl, url, req);
+    /**
+     * Make a REST call to the gatekeeper.
+     *
+     * @param {String} path The path to be added to the gatekeeper url
```
Comment (dijkstraj) [inline src/modules/sessionManager.js]:
```suggestion
     * @param {string} path The path to be added to the gatekeeper url
```

**Review by dijkstraj — CHANGES_REQUESTED** [review]

**Review by dijkstraj — APPROVED** [review]
